src/Advection/weno_fifth_order_stretched.jl

```diff
@@ -0,0 +1,339 @@
+#####
+##### Weighted Essentially Non-Oscillatory (WENO) scheme of order 5
+#####
+
+using OffsetArrays
+using Oceananigans.Grids: with_halo
+using Oceananigans.Architectures: arch_array
+using Adapt
+import Base: show
+
+const AG = AbstractGrid
+const two_32 = Int32(2)
+
+const C3₀ = 3/10
+const C3₁ = 3/5
+const C3₂ = 1/10
+
+const ƞ = Int32(2) # WENO exponent
+const ε = 1e-6
+
+struct WENO5{FT, XT, YT, ZT, Buffer} <: AbstractUpwindBiasedAdvectionScheme{2} 
+    
+    coeff_xᶠᵃᵃ::XT
+    coeff_xᶜᵃᵃ::XT
+    
+    coeff_yᵃᶠᵃ::YT
+    coeff_yᵃᶜᵃ::YT
+    
+    coeff_zᵃᵃᶠ::ZT
+    coeff_zᵃᵃᶜ::ZT
+ 
+end
+
+function WENO5(FT = Float64; grid = nothing) 
+    
+    if grid isa Nothing
+        coeff_xᶠᵃᵃ = nothing 
+        coeff_xᶜᵃᵃ = nothing  
+        coeff_yᵃᶠᵃ = nothing   
+        coeff_yᵃᶜᵃ = nothing
+        coeff_zᵃᵃᶠ = nothing
+        coeff_zᵃᵃᶜ = nothing
+    elseif !(typeof(grid) <: RectilinearGrid)
+        @warn "Stretched WENO is not supported with grids other than Rectilinear, defaulting to Uniform WENO"
+        coeff_xᶠᵃᵃ = nothing 
+        coeff_xᶜᵃᵃ = nothing  
+        coeff_yᵃᶠᵃ = nothing   
+        coeff_yᵃᶜᵃ = nothing
+        coeff_zᵃᵃᶠ = nothing
+        coeff_zᵃᵃᶜ = nothing    
+    else
+        FT          = eltype(grid)
+        arch        = grid.architecture
+        helper_grid = with_halo((4, 4, 4), grid)
+        if typeof(grid) <: XRegRectilinearGrid 
+            coeff_xᶠᵃᵃ = nothing 
+            coeff_xᶜᵃᵃ = nothing 
+        else
+            coeff_xᶠᵃᵃ = calc_interpolating_coefficients(FT, helper_grid.xᶠᵃᵃ, arch, grid.Nx) 
```

Comment:
Might be _slightly_ simpler to define 

```julia
calc_interpolating_coefficients(FT, ::XRegRectilinearGrid, arch, grid.Nx) = nothing
```

then you don't need the if statement. Same for below.

Reply:
Hmmm then you would need to define a separate `calc_interpolating_coefficients` for x-, y- and z-directions

Reply:
```julia
calc_interpolating_coefficients(FT, ::XRegRectilinearGrid, arch, grid.Nx) = nothing

calc_interpolating_coefficients(FT, ::XRegRectilinearGrid, arch, grid.Nx) = nothing

Reply:
how about

```julia
calc_interpolating_coefficients(FT, ::OffsetArray{<:Any, <:AbstractRange}, arch, grid.Nx) = nothing
```

Reply:
Sure, that works!